<SegmentedToggle/> - new component

mykas opened the pull request:
Shoutout composer initiative. 

Initial API.

```jsx
<SegmentedToggle defaultChecked="short" onChange={onChange}>
  <SegmentedToggle.Button value="short">Short Option</SegmentedToggle.Button>
  <SegmentedToggle.Button value="medium">Medium Option</SegmentedToggle.Button>
  <SegmentedToggle.Button value="long">Long Option</SegmentedToggle.Button>
<SegmentedToggle>
```

argshook commented on src/SegmentedToggle/ToggleButton/ToggleButton.js:
```diff
@@ -0,0 +1,31 @@
+import React from 'react';
```
would it ever be a valid case to use `ToggleButton` as standalone component not, and likely
`SegmentedToggle` is a compound component, meaning, ToggleButton should always be together with
`SegmentedToggle`.

if so, then i'd go with something like `SegmentedToggle.Button` and avoid
separate export/import.

mykas replied:
Moved to `<SegmentedToggle/>` composite.

argshook commented on src/SegmentedToggle/SegmentedToggle.js:
```diff
@@ -0,0 +1,36 @@
+  };
+
+  state = {
+    checked: this.props.initiallyChecked,
```
i understand this is a WIP but please don't forget about multiple options, cause this component
should support it.

it will likely look something very similar to what we already brainstormed with `SelectableList`

mykas replied:
Changed to "defaultChecked" and agreed to limit component to support only single selection.

argshook commented on src/SegmentedToggle/ToggleButton/ToggleButton.spec.js:
```diff
@@ -0,0 +1,53 @@
+    expect(await driver.exists()).toBeTruthy();
+  });
+
+  it('should be controlled', async () => {
```
controlled is that which is controlled from the outside, as in, has props passed
uncontrolled is that which is not controlled from outside, and controls it's own state itself

here, since state changes after click but outside world doesn't change, component is `uncontrolled`.

these are confusing as hell, i know

mykas replied:
But ToggleButton is controlled from outside and it doesn't have its own state.

argshook commented on src/SegmentedToggle/ToggleButton/ToggleButton.js:
```diff
@@ -0,0 +1,50 @@
+);
+
+ToggleButton.PropTypes = {
+  children: string,
```
it should be `node`, cause what if i want `<Translate key={my.translation.whatever}/>` as children?

rule of thumb is: when you think `string` for prop, you actually think `node`

mykas replied:
Oh yes yes..

argshook commented on src/SegmentedToggle/ToggleIcon/ToggleIcon.st.css:
```diff
@@ -0,0 +1,83 @@
+}
+
+
+.root {
```
how different is this file from `ToggleButton`? Maybe you can reuse stuff?

mykas replied:
welll I could use stylable mixins :trollface:

argshook commented on stories/SegmentedToggle/index.story.js:
```diff
@@ -0,0 +1,95 @@
+            "import SegmentedToggle from 'wix-style-react/SegmentedToggle';",
+        }),
+        description({
+          text: `🙋 Component includes compound components: Button and Icon.  With following components user can build 3 types of SegmentedToggles:`,
```
you can use markdown here, so no need for multiple `descriptions` just for new lines

mykas commented on src/SegmentedToggle/SegmentedToggle.driver.private.js:
```diff
@@ -0,0 +1,7 @@
+import { segmentedToggleDriverFactory as publicDriverFactory } from './SegmentedToggle.driver';
+
+export const segmentedTogglePrivateDriverFactory = base => {
```
Sure

mykas commented on src/SegmentedToggle/SegmentedToggle.spec.js:
```diff
@@ -0,0 +1,59 @@
+  );
+
+  describe('`onClick` handler should return', () => {
+    it('when clicked with a mouse', async () => {
```
No need, children are based on `button`.

mykas replied:
Added

mykas commented on src/SegmentedToggle/ToggleButton/ToggleButton.js:
```diff
@@ -0,0 +1,50 @@
+
+import Text from '../../Text';
+
+const _addPrefix = icon =>
```
true

mykas commented:
@LittleTal Hey, can you check https://wix-wix-style-react-pr-2797.surge.sh/?selectedKind=Components&selectedStory=SegmentedToggle&full=0&addons=0&stories=1&panelRight=0 ? and

mykas commented:
Also, check these test pages which shows behaviour: https://wix-wix-style-react-pr-2797.surge.sh/?selectedKind=Tests%2FComponents%2FSegmentedToggle&selectedStory=Text%20%26%20Prefix&full=0&addons=0&stories=1&panelRight=0 and http://localhost:6006/?selectedKind=Tests%2FComponents%2FSegmentedToggle&selectedStory=Icon&full=0&addons=0&stories=1&panelRight=0

mykas commented:
thanks! 

* Fixed.
* You gonna need to show me this case. 
* Fixed. 
* Fixed.
* I am not sure if this is possible right now, because text ellipsis is bound to text and in your defined case I need to extract ellipsis from Text component and detect when text gets too little space that it gets `...`. But you can open an issue or feature request for this.

LittleTal commented:
@mykas you should keep the hover effect, only without covering the toggle border while hovering

argshook commented on src/SegmentedToggle/SegmentedToggle.js:
```diff
@@ -0,0 +1,65 @@
+        {React.Children.map(children, (child, index) =>
+          React.cloneElement(child, {
+            disabled,
+            'data-click': `segmented-toggle-${index + 1}`,
```
use `data-hook`

mykas replied:
This case is different. I need to change how data-hook is passed in order to be able to test it in SegmentedToggle with method selectChild. By using custom `data` attribute I don't need to reorder how ToggleIcon handles data-hook.

argshook replied:
what i mean is that `ToggleButton` and `ToggleIcon` should both support dataHook and you should use
them to target elements in tests. That's exactly what dataHook is for, how is new name
(`data-click`) helping?

argshook replied:
also, why `+ 1`?

mykas replied:
ok, let me explain. ToggleIcon can be clicked only if I would use testkit which handles click handler. The reason for this is because Popover wraps the `button` component. So simple base.click() would not just work.

I want to expose for users selectChild() method in SegmentedToggle where they could click either ToggleButton or ToggleIcon. So here is the problem: how should I handle click method inside SegmentedToggle.driver if one can be clicked like this: base.click() and another one needs to be used with testkit? this data attribute way was the easiest for me to implement without changing the code.

P.S. `+1` seems to be easier for users to selectChild. It is documented in api docs too.

argshook commented on src/SegmentedToggle/SegmentedToggle.js:
```diff
@@ -0,0 +1,65 @@
+    const { onClick } = this.props;
+    const { value } = evt.currentTarget;
+    this.setState({ selected: value }, () =>
+      onClick ? onClick(evt, value) : null,
```
why `null`?
```suggestion
      onClick && onClick(evt, value),
```

argshook commented on src/SegmentedToggle/ToggleIcon/ToggleIcon.js:
```diff
@@ -0,0 +1,83 @@
+
+  open = () => {
+    const { disabled } = this.props;
+    this.setState({ shown: disabled ? false : true });
```
```suggestion
    this.setState({ shown: !disabled });
```

argshook commented on src/SegmentedToggle/ToggleIcon/ToggleIcon.js:
```diff
@@ -0,0 +1,83 @@
+  static displayName = 'SegmentedToggle.Icon';
+
+  state = {
+    shown: false,
```
what `shown`? maybe `tooltipVisible` or something?

mykas replied:
its a copy paste from Yardens example in Popover :trollface:

argshook commented on src/SegmentedToggle/SegmentedToggle.driver.js:
```diff
@@ -0,0 +1,11 @@
+    element: async () => await base.getNative(),
+    /* selects child by given number (selection begins with 1) */
+    selectChild: async number =>
+      base.$(`[data-click="segmented-toggle-${number}"]`).click(),
```
https://github.com/wix/wix-style-react/pull/2797/files#r252577365

```suggestion
      base.$(`[data-hook="segmented-toggle-${number + 1}"]`).click(),
```
